src/dromajo_manycore.cpp

```diff
@@ -0,0 +1,314 @@
+    }
+    break;
+  }
+  return empty;
+}
+
+/*
+ * Write to the desired FIFO
+ * @param[in] type - Type of FIFO to write
+ * @param[in] offset - Chooses which 32-bit FIFO to write
+ * @param[in] val - The 32-bit value to be written
+ * @return FIFO write status
+ *  false --> Write was unsuccesful
+ *  true --> Write was successful
+ */
+bool mc_fifo_write(mc_fifo_type_t type, uint32_t fifo_id, uint32_t val) {
+  bool fifo_write_success = false;
+  switch (type) {
+    case FIFO_HOST_TO_MC_REQ:
+      fifo_write_success = fifo_write(host_to_mc_req_fifo, fifo_id, val);
+    break;
+    case FIFO_MC_TO_HOST_REQ:
+      fifo_write_success = fifo_write(mc_to_host_req_fifo, fifo_id, val);
+    break;
+    case FIFO_MC_TO_HOST_RESP:
+      fifo_write_success = fifo_write(mc_to_host_resp_fifo, fifo_id, val);
+    break;
+    default:
+    {
+      printf("Undefined/Unknown FIFO type!\nExiting...\n");
+      exit(-1);
+    }
+    break;
+  }
+  return fifo_write_success;
+}
+
+/*
+ * Read from the desired FIFO
+ * @param[in] offset - Chooses which 32-bit FIFO to read
+ * @param[in] val - Pointer to store the 32-bit read value 
+ * @param[in] type - Type of FIFO to read
+ * @return FIFO read status
+ *  false --> Read was unsuccesful
+ *  true --> Read was successful
+ */
+bool mc_fifo_read(mc_fifo_type_t type, uint32_t fifo_id, uint32_t *val) {
+  bool fifo_read_success = false;
+  switch (type) {
+    case FIFO_HOST_TO_MC_REQ:
+      fifo_read_success = fifo_read(host_to_mc_req_fifo, fifo_id, val);
+    break;
+    case FIFO_MC_TO_HOST_REQ:
+      fifo_read_success = fifo_read(mc_to_host_req_fifo, fifo_id, val);
+    break;
+    case FIFO_MC_TO_HOST_RESP:
+      fifo_read_success = fifo_read(mc_to_host_resp_fifo, fifo_id, val);
+    break;
+    default:
+    {
+      printf("Undefined/Unknown FIFO type!\nExiting...\n");
+      exit(-1);
+    }
+    break;
+  }
+  return fifo_read_success;
+}
+
+/*
+ * Get the credits for the desired FIFO
+ * Only BP will request the manycore for credits
+ * @params[in] type - Type of FIFO requesting credits
+ * @returns the remaining FIFO credits
+ */
+int mc_fifo_get_credits(mc_fifo_type_t type) {
+  int credits = 0;
+  switch (type) {
+    case FIFO_HOST_TO_MC_REQ:
+      credits = get_fifo_credits(host_to_mc_req_fifo);
+    break;
+    case FIFO_MC_TO_HOST_REQ:
+    case FIFO_MC_TO_HOST_RESP:
+      printf("Manycore will not request credits from the host\n");
+    break;
+    default:
+    {
+      printf("Undefined/Unknown FIFO type!\nExiting...\n");
+      exit(-1);
+    }
+    break;
+  }
+  return credits;
+}
+
+/*
+ * Returns if a given 32-bit or 128-bit FIFO is empty
+ * @params[in] fifo - 128-bit FIFO to check emptiness
+ * @params[in] _empty - Set to true if you need the emptiness of a given 32-bit FIFO
+ * @params[in] fifo_id - Chooses which 32-bit FIFO's emptiness to return
+ * @returns 32-bit/128-bit FIFO emptiness
+ */
+bool get_fifo_empty(mc_fifo_t *fifo, bool _empty, uint32_t fifo_id) {
```

Comment:
How is this used? Primarily we should be querying credits, not emptiness

Reply:
I use the empty function to check is there are packets to transmit. It is also used on the receiving FIFOs if there are any entries. Credits is a different mechanism altogether. The overarching fifo struct has a field to maintain credits that gets updated by the x86 code after every packet transmission.

Reply:
Okay, so this is read by the manycore_simulator, not the software ever?

Reply:
The software never uses this function directly. It doesn't even know this function exists. This exists only for Dromajo to know how to handle the device. And strictly speaking not even the simulator uses this particular function. The simulator uses the mc_get_fifo_empty function which calls this function internally.

To be exact this is what happens under different cases -->
Software loads from the FIFO address --> mc_fifo_read is called on the correct FIFO
Software stores to the FIFO address --> mc_fifo_write is called on the correct FIFO
Software wants to know the number of credits used --> mc_fifo_get_credits is called on the FIFO (prints a warning message if this function is triggerred)
Software wants to check the if entries present in the FIFO --> mc_is_fifo_empty called on the correct FIFO (The function is generic but always the software checks for emptiness of the manycore request/response FIFOs and the simulator checks the emptiness of the host request FIFO)